daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/EngineTest.scala

```diff
@@ -1570,6 +1539,104 @@ class EngineTest
         )
       }
     }
+
+    "unused disclosed contracts not saved to ledger" in {
+      val templateId = Identifier(basicTestsPkgId, "BasicTests:WithKey")
+      val usedContractSKey = SValue.SRecord(
+        templateId,
+        ImmArray("_1", "_2").map(Ref.Name.assertFromString),
+        values = ArrayList(SValue.SParty(alice), SValue.SInt64(42)),
+      )
+      val usedContractKey = Value.ValueRecord(
+        None,
+        ImmArray(
+          None -> Value.ValueParty(alice),
+          None -> Value.ValueInt64(42),
+        ),
+      )
+      val unusedContractKey = Value.ValueRecord(
+        None,
+        ImmArray(
+          None -> Value.ValueParty(alice),
+          None -> Value.ValueInt64(69),
+        ),
+      )
+      val usedDisclosedContract = DisclosedContract(
+        templateId,
+        SValue.SContractId(toContractId("BasicTests:WithKey:1")),
+        SValue.SRecord(
+          templateId,
+          ImmArray(Ref.Name.assertFromString("p"), Ref.Name.assertFromString("k")),
+          ArrayList(SValue.SParty(alice), SValue.SInt64(42)),
+        ),
+        ContractMetadata(
+          now,
+          Some(crypto.Hash.assertHashContractKey(templateId, usedContractKey)),
+          ImmArray.empty,
+        ),
+      )
+      val unusedDisclosedContract = DisclosedContract(
+        templateId,
+        SValue.SContractId(toContractId("BasicTests:WithKey:2")),
+        SValue.SRecord(
+          templateId,
+          ImmArray(Ref.Name.assertFromString("p"), Ref.Name.assertFromString("k")),
+          ArrayList(SValue.SParty(alice), SValue.SInt64(42)),
+        ),
+        ContractMetadata(
+          now,
+          Some(crypto.Hash.assertHashContractKey(templateId, unusedContractKey)),
+          ImmArray.empty,
+        ),
+      )
+      val lookupByKeyCommand = speedy.Command.LookupByKey(
+        templateId = templateId,
+        contractKey = usedContractSKey,
+      )
+
+      unusedDisclosedContractsNotSavedToLedger(
+        lookupByKeyCommand,
+        unusedDisclosedContract,
+        usedDisclosedContract,
+      )
+    }
+  }
+
+  "fetch template" should {
```

Comment:
`FetchTemplate` testing - needs to be reviewed

Reply:
nit: it’s fetch by contract id not fetch by template.

Reply:
Here I was referring to `speedy.command.FetchTemplate`.